### PR TITLE
Pvinis master

### PR DESCRIPTION
**Pull request by pvinis**
## Description

There are two things here.
- One new song
- A bunch of changes on my files, for sharing things between my keyboards.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/ergotravel/keymaps/pvinis/keymap.c [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+
+
+
+
+
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    case QWERTY:
```
Comment (noroadsleft) [inline keyboards/ergotravel/keymaps/pvinis/keymap.c]: Lines 92-100 aren't being used in your keymap at all. They can be deleted.

File: keyboards/iris/keymaps/pvinis/config.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include "config_common.h"
```
Comment (noroadsleft) [inline keyboards/iris/keymaps/pvinis/config.h]: This line is redundant.

```suggestion
```

File: keyboards/minidox/keymaps/pvinis/keymap.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+#include QMK_KEYBOARD_H
+#include "pvinis.h"
+
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+#define _QWERTY 0
```
Comment (noroadsleft) [inline keyboards/minidox/keymaps/pvinis/keymap.c]: Lines 24-27 would be better as an enum:

```c
enum layers {
    _QWERTY,
    _LOWER,
    _RAISE,
    _ADJUST,
};
```

File: keyboards/minidox/keymaps/pvinis/keymap.c [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+#define _ADJUST 16
+
+enum custom_keycodes {
+  QWERTY = SAFE_RANGE,
+  LOWER,
+  RAISE,
+  ADJUST,
+};
+
+// Fillers to make layering more clear
```
Comment (noroadsleft) [inline keyboards/minidox/keymaps/pvinis/keymap.c]: Lines 36-38 should be deleted. QMK defines these keycode aliases automatically.

https://github.com/qmk/qmk_firmware/blob/a120a000d481ed14a87988383be46c5a76604bb6/quantum/quantum_keycodes.h#L29-L31

File: keyboards/minidox/keymaps/pvinis/keymap.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+};
+
+void persistant_default_layer_set(uint16_t default_layer) {
+  eeconfig_update_default_layer(default_layer);
+  default_layer_set(default_layer);
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    case QWERTY:
```
Comment (noroadsleft) [inline keyboards/minidox/keymaps/pvinis/keymap.c]: Lines 142-150 aren't being used at all and can be deleted.

File: keyboards/minidox/keymaps/pvinis/keymap.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+ */
+[_ADJUST] =  LAYOUT( \
+  KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,        KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10, \
+  KC_F11,  KC_F12,  _______, _______, _______,      _______, _______, _______, TSKMGR, CALTDEL, \
+  RESET,   _______, _______, _______, _______,      _______, _______, _______, _______,  _______, \
+                    _______, _______, _______,      _______,  _______, _______                    \
+)
+};
+
+void persistant_default_layer_set(uint16_t default_layer) {
```
Comment (drashna) [inline keyboards/minidox/keymaps/pvinis/keymap.c]: Could you remove this function?  

It's been depreciated in favor of the `set_single_persistent_default_layer` function (which is a core function).

https://docs.qmk.fm/#/ref_functions?id=setting-the-persistent-default-layer

File: users/pvinis/pvinis.h [hunk 7]
```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+#include "quantum.h"
+
+
+// my own keycodes
+enum userspace_custom_keycodes {
+  PV_ = SAFE_RANGE,
```
Comment (drashna) [inline users/pvinis/pvinis.h]: This looks like you're using a placeholder.  You don't actually need to do that.  So this should be `PV_VRSN`, instead.

File: users/pvinis/rules.mk [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+AUDIO_ENABLE     = no  # piezo speaker sounds
+RGBLIGHT_ENABLE  = no  # rgb leds underlight
+TAP_DANCE_ENABLE = yes
+BACKLIGHT_ENABLE = no  # leds under keycaps
+#MOUSEKEY_ENABLE  = no
+#SLEEP_LED_ENABLE = no  # no led blinking while sleeping
+#NKRO_ENABLE      = yes
+
+# make firmware smaller
+EXTRAFLAGS += -flto
```
Comment (drashna) [inline users/pvinis/rules.mk]: This should be replaced, actually:
```suggestion
LINK_TIME_OPTIMIZATION_ENABLE = yes
```

This also disables "ACTION_MACRO" and "ACTION_FUNCTION" features, which are depreciated.

File: users/pvinis/keymap.c [hunk 9]
```diff
@@ -0,0 +1,122 @@
+// pvinis iris
```
Comment (drashna) [inline users/pvinis/keymap.c]: I think this file is supposed to go into `/keyboards/keebio/iris/keymaps/`, not here.

File: quantum/audio/song_list.h [hunk 10]
```diff
@@ -503,6 +502,17 @@
     W__NOTE(_C5),      \
     W__NOTE(_BF4),
 
+#define KATAMARI_ROLLING_STAR \
```
Comment (drashna) [inline quantum/audio/song_list.h]: For now, could this be left out?

File: keyboards/ergodox_ez/keymaps/pvinis/rules.mk [hunk 11]
```diff
@@ -0,0 +1,6 @@
+TAP_DANCE_ENABLE = yes
+
+
+ifndef QUANTUM_DIR
```
[the remark this answers is not in the document]
Comment (pvinis) [reply keyboards/ergodox_ez/keymaps/pvinis/rules.mk]: oh. is it part of the qmk now?

[the remark this answers is not in the document]
Comment (drashna) [reply keyboards/ergodox_ez/keymaps/pvinis/rules.mk]: It's a holdover from the old make system that we used to use. 

Now, all it does it potentially create compiler warnings.

File: users/pvinis/pvinis.h (hunk 7 above)
Comment (pvinis) [reply users/pvinis/pvinis.h]: I do use it as a placeholder. I think it's nice, one missed keycode should be ok to be used like that, right?

File: quantum/audio/song_list.h (hunk 10 above)
Comment (pvinis) [reply quantum/audio/song_list.h]: Should I put this in another PR?

Comment (pvinis) [reply quantum/audio/song_list.h]: Moved to https://github.com/qmk/qmk_firmware/pull/5854

**Comment by pvinis** [comment]
I removed some keymaps as they were just a copy of the default ones. I'll add them here later, when I make them by using the userspace code.

**Comment by pvinis** [comment]
@drashna I have one question before we can merge this. I made this `keyboard_post_init_user_local` that I can put keyboard-keymap specific code there, like the led sequence on the ez, but I don't need it on the iris.
Do you have an idea of how to have a "default" empty function somewhere and if a keymap has it, it;s overridden?

**Comment by pvinis** [comment]
you didn't answer my question :(

do you have any ideas on how to do that?

**Comment by drashna** [comment]
Oh, sorry.   

And that is actually covered in the docs:
https://docs.qmk.fm/#/feature_userspace?id=customized-functions

Specifically, you want the `__attribute__ ((weak))` line before the function.  This tells the compiler that this function can be replaced by code in another file.

**Comment by pvinis** [comment]
awesome. I checked that doc but I missed that somehow 😬

thanks!

**Comment by drashna** [comment]
Well, there are a lot of docs, and a lot of information in them, so I can totally understand that!

**Pull request merged**